File: cpp/93_Restore_IP_Addresses.cpp
/*
Given a string containing only digits, restore it by returning all possible valid IP address combinations.

Example:

Input: "25525511135"
Output: ["255.255.11.135", "255.255.111.35"]
*/
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        vector<string> restoreIpAddresses(string s) {
            if(s.size() > 12) return {};
            vector<string> res;
            string comp = "";
            backtracking(res, s, comp, 0, 0);
            return res;
        }

        void backtracking(vector<string> &res, string s, string comp, int index, int count) {
            if(count > 4) return;
            if(index == s.size() && count == 4) {
                res.push_back(comp);
            }
            for(int i = 1; i < 4; i++) {
                if(index + i > s.size()) break;
                string t = s.substr(index, i);
                if((t[0] == '0' && t.size() > 1) || stoi(t) > 255) continue;
                backtracking(res, s, comp + t + (count == 3 ? "" : "."), index + i, count + 1);
            }
        }
};

int main() {
    Solution sol;
    return 0;
}

/*


*/

File: py/93_Restore_IP_Addresses.py
from typing import List

class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        if len(s) > 12:
            return []
        res = []
        self.backtracking(s, [], res)
        return res
    
    def backtracking(self, remain: str, ipstr: List[str], res: List[str]):
        if len(ipstr) > 0 and (int(ipstr[-1]) > 255 or (ipstr[-1][0] == '0' and len(ipstr[-1]) > 1)):
            return
        if len(remain) == 0 and len(ipstr) == 4:
            res.append('.'.join(ipstr))
            return
        for i in range(1, min(len(remain) + 1, 4)):
            self.backtracking(remain[i:], ipstr + [remain[0:i]], res)
        